forc-plugins/forc-crypto/src/keys/vanity.rs

```diff
@@ -0,0 +1,440 @@
+    }
+
+    println!("Starting to generate vanity address...");
+    let start_time = Instant::now();
+
+    let (address, secret_key, mnemonic) =
+        match (left_exact_hex, left_regex, right_exact_hex, right_regex) {
+            (Some(left), _, Some(right), _) => {
+                let matcher = HexMatcher { left, right };
+                find_vanity_address(matcher, mnemonic)
+            }
+            (Some(left), _, _, Some(right)) => {
+                let matcher = LeftExactRightRegexMatcher { left, right };
+                find_vanity_address(matcher, mnemonic)
+            }
+            (_, Some(left), _, Some(right)) => {
+                let matcher = RegexMatcher { left, right };
+                find_vanity_address(matcher, mnemonic)
+            }
+            (_, Some(left), Some(right), _) => {
+                let matcher = LeftRegexRightExactMatcher { left, right };
+                find_vanity_address(matcher, mnemonic)
+            }
+            (Some(left), None, None, None) => {
+                let matcher = LeftHexMatcher { left };
+                find_vanity_address(matcher, mnemonic)
+            }
+            (None, None, Some(right), None) => {
+                let matcher = RightHexMatcher { right };
+                find_vanity_address(matcher, mnemonic)
+            }
+            (None, Some(re), None, None) => {
+                let matcher = SingleRegexMatcher { re };
+                find_vanity_address(matcher, mnemonic)
+            }
+            (None, None, None, Some(re)) => {
+                let matcher = SingleRegexMatcher { re };
+                find_vanity_address(matcher, mnemonic)
+            }
+            _ => return Err(anyhow!("Invalid pattern combination")),
+        }?;
+
+    let duration = start_time.elapsed();
+    println!(
+        "Successfully found vanity address in {:.3} seconds.",
+        duration.as_secs_f64()
+    );
+
+    let result = if let Some(mnemonic) = mnemonic {
+        json!({
+            "Address": address.to_string(),
+            "PrivateKey": hex::encode(secret_key.as_ref()),
+            "Mnemonic": mnemonic,
+        })
+    } else {
+        json!({
+            "Address": address.to_string(),
+            "PrivateKey": hex::encode(secret_key.as_ref()),
+        })
+    };
+
+    if let Some(path) = save_path {
+        fs::write(path, serde_json::to_string_pretty(&result)?)?;
+    }
+
+    Ok(result)
+}
+
+/// Generates random wallets until `matcher` matches the wallet address, returning the wallet.
+pub fn find_vanity_address<T: VanityMatcher>(
+    matcher: T,
+    use_mnemonic: bool,
+) -> Result<(Address, SecretKey, Option<String>)> {
+    wallet_generator(use_mnemonic)
+        .find_any(|result| {
+            if let Ok((addr, _, _)) = result {
+                matcher.is_match(addr)
+            } else {
+                false
+            }
+        })
+        .ok_or_else(|| anyhow!("Failed to generate a matching address"))?
+}
```

Comment:
What are your thoughts about having some user feedback? something like...

```rust
pub fn find_vanity_address<T: VanityMatcher>(
    matcher: T, 
    use_mnemonic: bool,
) -> Result<(Address, SecretKey, Option<String>)> {
    let start = Instant::now();
    let mut attempts = AtomicUsize::new(0); // Need atomic for parallel iteration
    
    let result = wallet_generator(use_mnemonic)
        .for_each(|_| {
            let current = attempts.fetch_add(1, Ordering::Relaxed);
            if current % 10000 == 0 { // Pick a nice number that makes sense here, just going with 10000 
                let elapsed = start.elapsed().as_secs_f64();
                let rate = current as f64 / elapsed;
                println!("Tried {} addresses ({:.2} addresses/sec)...", current, rate);
            }
        })
        .find_any(|result| {
            if let Ok((addr, _, _)) = result {
                matcher.is_match(addr)
            } else {
                false
            }
        })
        .ok_or_else(|| anyhow!("Failed to generate a matching address"))?;

    Ok(result?)
}
```

Reply:
this can be a good idea; the `solana-keygen` seems to do this.

Reply:
addressed: https://github.com/FuelLabs/sway/pull/6661/commits/e493891bc22b376d7175321e1367b2342587896f